### DISCUSSION TITLE
Some beginner feedback

### DISCUSSION THREAD

**Discussion by quat1024** [nushell/nushell#10670] [General]
Hi, i started messing with Nushell today and I'm finding it very compelling. Today I ran `optipng` on a directory tree full of png files and parallelizing it was super easy thanks to `par-each`. It's nice not having to look up a bunch of bespoke "ok how do i sort by filesize" flags because if I can see it I can `sort-by` it... and if I can see it I can also pipe it to a `.nuon` file, which opened up some interesting possibilities. I saved my `ls` output to a file before compressing the images, then all it took was a `load` and a `join` to start answering questions like "what images got compressed the most" and "how much space did optipng save in total", and without nushell making it easy I don't think I would have even bothered to ask the questions. I never liked shell scripting but maybe it *can* be fun, hm.

Anyway, as I learned more I started taking some notes about things I got stuck on.

### Tables vs records vs lists

Records are collections keyed by name, lists are collections keyed by sequential integer, cool. Sounds like tables can be understood as a *list of records* (where each record corresponds to a row) or as a *record of lists* (where each list corresponds to a column). I can pluck out individual columns with `get <column name>` which returns a list, or rows with `get <row id>` which returns a record.

But many of the other operations feel... asymmetrical?

* I can convert lists to one-column tables with `wrap`, but for the life of me I can't find a good way to convert a record to a one-*row* table, which is weird because records already *are* conceptual table rows. Well I came up with `| transpose | transpose | reject column0 | headers` and `| (let in2 = $in; [[];[]] | append $in2 | skip)` anyway but they both suck.
* I can convert records to lists with `| values`, but I can't find a good way to convert a list to a record. There's `| into record` but I can't easily supply the column names. I expected `[1 2 3] | transpose col1 col2 col3` to produce a one-row table (which i can slice into a record with `| get 0`), but `transpose` doesn't work on lists for some reason (even though it works on records) and I need to do `[1 2 3] | wrap "dummy" | transpose -i col1 col2 col3`...
* `append` is one command for adding records as a new table row, but adding lists as a new table column needs two (`wrap` and `merge`). `insert` is for something else.
* I can change every row with `each` and change one column with `update`, but I don't know if theres an easy way to change *one* row (like an indexed `append`) or change *every* column (like a columnar `each`... maybe not so useful).
* On [the filters page](http://www.nushell.sh/commands/categories/filters.html) I saw `drop` and `drop column`, and thought oh, maybe there are `column`/`row`-suffixed versions of other commands! But actually I think `drop column` is the only one lol

I understand that table operations being "symmetrical" isn't a goal, because columns and rows have different functions. And maybe converting a record to a single-row table is silly, although sometimes I want to do it just to make the pretty-printer not display it vertically 😩

But still, in order to use nushell effectively I want to understand the primitive operations on tables/lists/records so I can start thinking in terms of them, and if they're lopsided it's kind of frustrating. Going one way is easy but the other way is hard...

### Docs

I was looking around for "all the functions that operate on tables" to fill in some of the gaps in my understanding mentioned above, and it took me a while to find the "filters" section. There's a lot of good functions in there but I don't know if there are others I'd be interested in that aren't counted as "filters", and I don't really know what a filter is anyway (given that apparently `scope aliases` counts as one)

Some commands (like `query xml`) are actually part of non-default plugins but are mixed in with the rest of the commands and the documentation doesn't mention they're part of a plugin. Also some of the examples are kinda underwhelming. I think i'm interested in contributing to the docs if i have some free time ^^

### The pretty printer

The pretty printer is just too damn smart for me lol!

* Lists are printed vertically, which makes sense because they correspond to table columns. But so are records, even though they're table rows. I noticed the ["Working with tables"](http://www.nushell.sh/book/working_with_tables.html#adding-a-new-column) book chapter alludes to the automatic rotation being somewhat unexpected behavior (and also hints at one-row tables being printed vertically too? ugh):
    > This visual may be slightly confusing, because it looks like what we've just done is add a row. In this case, remember: rows have numbers, columns have names. If it still is confusing, note that appending one more row will make the table **render as expected:**
* If a table has a column named `index` it gets printed as `#`. This was very surprising behavior for me. I was trying [the first example for the `merge` command](http://www.nushell.sh/commands/docs/merge.html#examples) and it still looked like a one-column table. Got halfway through opening a "merge is broke!!!11" bug report before I noticed the `#` column was off by one...

I don't think there's a way to turn off these pretty-print features, which would be really nice for learning.

**Comment by fdncred** [comment]
Thanks for your detailed feedback. Nushell moves fast but we definitely would like more consistency like you point out. Here's some points that may help.

> tables, records, lists

This is the definition we usually share. You can learn more here https://www.nushell.sh/book/types_of_data.html#types-of-data
* a list is a collection of items with no key
* a record holds at least 1 key value pair
* a table is a list of records

> record to table

I generally agree that we need better record commands and `into record` needs some love.

record
```nushell
❯ {key1: val1 key2: val2}
╭────┬────╮
│key1│val1│
│key2│val2│
╰────┴────╯
```
record turned into a table
```nushell
❯ [{key1: val1 key2: val2}]
╭#┬key1┬key2╮
│0│val1│val2│
╰─┴────┴────╯
```

> list to record

This is a little weird because there isn't a straight conversion since you have to supply data, i.e., the keys do not already exist. So, saying there is no conversion is right, but we may be able to generate something.

Here's one way. Since `[1 2 3]` is a list, you can do this to produce a list of records aka a table. 
```nushell
❯ [1 2 3] | enumerate
╭#┬item╮
│0│   1│
│1│   2│
│2│   3│
╰─┴────╯
```
The `enumerate` command creates a list of items with an `index` and an `item`. Since any table with the column name `index` is turned into the tables index, we see those listed under `#` and the items in the `item` column.

> update a row

This could be a recently introduced bug because you should be able to do this to insert or update at the row index. (update: I tested this and it still works on the latest main).
```nushell
ls | update 3 {name: I type: love size: nushell modified: always }
```
The key with the `update 3` example is that the data needs to match the shape of the item you're updating. Since we're updating a table, which is a list of records, I create the update in the shape of a record with the exact keys as column names.

> drop and drop column

Agreed that this is confusing. We have `drop nth` that maybe could be named `drop row`

```nushell
ls | drop nth 3
```

> docs and help

If you haven't discovered it yet `help -f term` is your friend. It helps you find commands related to the term you specify. It searches command names and search_items that are listed in `help commands`.

We could definitely do better with our documentation and are always looking for help to improve on them.

> tables

You can definitely turn off the index in tables. It's in the config.nu under table -> index_mode
